Create method configs for Dockstore, Agora [GAWB-3100]

aednichols opened the pull request:
To try this out, you can spin up UI -> Orch -> Front Rawls and use the APIs below.

Dockstore configs show up on the Method Configurations page and look only mildly broken.

Of course, you can't actually do anything with them until GAWB-3101.

To use the new functionality from Orch, https://github.com/broadinstitute/firecloud-orchestration/pull/607 is required; but since these changes are strictly additive, this PR can safely be merged first (i.e. everything that is documented and currently works in Orch, will keep working).

Still needs
- [ ] Signoff on the changes to the API relative to the tech doc!
- [ ] Update tech doc
- [x] Fix existing tests
- [x] Tests for the new things
- [ ] Update Rawls Swagger
- [ ] Update Orch Swagger
- [x] Small [UI PR](https://github.com/broadinstitute/firecloud-ui/pull/1180) to prettify Method Configurations Page

---

`POST /api/workspaces/{workspaceNamespace}/{workspaceName}/methodconfigs` now accepts and does the right thing with any of
```
"methodRepoMethod": { // Agora - backwards compatible
  "methodNamespace": "adam-methods",
  "methodName": "test",
   "methodVersion": 555
}
```
```
"methodRepoMethod": {
  "sourceRepo": "agora",
  "methodNamespace": "adam-methods",
  "methodName": "test",
   "methodVersion": 555
}
```
```
"methodRepoMethod": {
  "sourceRepo": "dockstore",
  "methodPath": "broadinstitute/wdl/Validate-Bams",
  "methodVersion": "develop"
}
```
```
"methodRepoMethod": {
  "methodUri": "dockstore://broadinstitute%2Fwdl%2FValidate-Bams/develop",
}
```
```
"methodRepoMethod": {
  "methodUri": "agora://adam-methods/test/555",
}
```

`GET /api/workspaces/{workspaceNamespace}/{workspaceName}/methodconfigs` now returns
```
// All existing fields plus two new ones
"methodRepoMethod": {
  "sourceRepo": "agora",
  "methodUri": "agora://adam-methods/test/555",
  "methodNamespace": "adam-methods", 
  "methodName": "test",
  "methodVersion": 555
}
```
for Agora MC's, and never returns any Dockstore configs.

`GET /api/workspaces/v2/{workspaceNamespace}/{workspaceName}/methodconfigs` is new and returns both Agora and Dockstore configs, like
```
"methodRepoMethod": {
  "sourceRepo": "agora",
  "methodUri": "agora://adam-methods/test/555",
  "methodNamespace": "adam-methods", 
  "methodName": "test",
  "methodVersion": 555
}
```
and
```
"methodRepoMethod": {
  "sourceRepo": "dockstore",
  "methodUri": "dockstore://broadinstitute%2Fwdl%2FValidate-Bams/develop",
  "methodPath": "broadinstitute/wdl/Validate-Bams",
  "methodVersion": "develop"
}
```

---

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

aednichols committed:
Clean model changes

aednichols committed:
New models; fixed existing tests

aednichols committed:
Cleanup

aednichols committed:
Cleanup

aednichols committed:
Comments

aednichols commented on model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala:
```diff
@@ -193,65 +193,133 @@ case class EntityCopyDefinition(
                    entityNames: Seq[String]
                    )
 
-// In the next phase (GAWB-3100), this class's members will be swapped out for a singular methodUri
```
In hindsight this seems wrong because referencing the fields of case classes is amazing, while on the other hand repeatedly parsing & re-generating a URI string in memory is painful

aednichols committed:
Better compatibility, no sourceRepo

aednichols committed:
Comment

aednichols committed:
It compiles

aednichols committed:
Fix hilarious JSON structure bug

kyuksel commented on core/src/main/scala/org/broadinstitute/dsde/rawls/util/WithSupport.scala:
```diff
@@ -27,11 +27,15 @@ trait MethodWiths {
+          case Success(Some(agoraEntity)) => op(agoraEntity)
+          case Failure(throwable) => DBIO.failed(new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.BadGateway, s"Unable to query the method repo.", methodRepoDAO.toErrorReport(throwable))))
+        }
+      case _ => throw new RawlsException("not yet supported")
```
Consider changing the message to 
```scala
s"$method not yet supported"
```

kyuksel commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/MethodConfigurationComponent.scala:
```diff
@@ -280,15 +280,15 @@ trait MethodConfigurationComponent {
 
     private def marshalMethodConfig(workspaceId: UUID, methodConfig: MethodConfiguration) = {
-      MethodConfigurationRecord(0, methodConfig.namespace, methodConfig.name, workspaceId, methodConfig.rootEntityType, methodConfig.methodRepoMethod.asAgoraMethodUrl, methodConfig.methodConfigVersion, methodConfig.deleted, methodConfig.deletedDate.map( d => new Timestamp(d.getMillis)))
+      MethodConfigurationRecord(0, methodConfig.namespace, methodConfig.name, workspaceId, methodConfig.rootEntityType, methodConfig.methodRepoMethod.methodUri, methodConfig.methodConfigVersion, methodConfig.deleted, methodConfig.deletedDate.map( d => new Timestamp(d.getMillis)))
```
Nit: I wonder if `method` has to be in so many places such as in `methodConfig.methodRepoMethod.methodUri`. For instance, are there `method`s that aren't `MethodRepoMethod`s here or `config`s that aren't `MethodConfig`s?

aednichols replied:
Fair point, but I kinda don't want to change it in the few places I'm touching without also changing All of the Things

kyuksel commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/MethodConfigurationComponent.scala:
```diff
@@ -280,15 +280,15 @@ trait MethodConfigurationComponent {
 
     private def unmarshalMethodConfigToShort(methodConfigRec: MethodConfigurationRecord): MethodConfigurationShort = {
-      MethodConfigurationShort(methodConfigRec.name, methodConfigRec.rootEntityType, MethodRepoMethod.apply(methodConfigRec.methodUri), methodConfigRec.namespace)
+      MethodConfigurationShort(methodConfigRec.name, methodConfigRec.rootEntityType, MethodRepoMethod.fromUri(methodConfigRec.methodUri), methodConfigRec.namespace)
```
glad `fromUri` worked out 👍

kyuksel commented on model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala:
```diff
@@ -479,7 +547,73 @@ class WorkspaceJsonSupport extends JsonSupport {
+      method match {
+        case agora: AgoraMethod =>
+          JsObject(agora.toJson.asJsObject.fields + ("methodUri" -> JsString(agora.methodUri)))
+        case dockstore: DockstoreMethod => JsObject("methodUri" -> JsString(dockstore.methodUri))
```
Move `JsObject(...)` to the line below as in the other `case`

kyuksel reviewed (commented):
It seems like it's coming along nicely :)

aednichols committed:
Code review tweaks

aednichols committed:
Code review tweaks

aednichols commented on core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala:
```diff
@@ -180,10 +180,14 @@ trait WorkflowSubmission extends FutureSupport with LazyLogging with MethodWiths
-      withWdl(method) { wdl =>
-        DBIO.successful(wdl)
-      }
+    methodConfig.methodRepoMethod match {
```
This change is not necessary

aednichols committed:
Code review tweaks

aednichols committed:
Fix

aednichols committed:
Hackish means of making tests compile

aednichols committed:
Possibly premature, can revert

aednichols committed:
Make everyone happy?

aednichols committed:
Cleanup

aednichols committed:
Tests for new stuff

aednichols committed:
Tests++

aednichols committed:
Merge branch 'develop' into an-GAWB-3100-v2

aednichols committed:
I love merge conflicts and RSI

aednichols committed:
Merge branch 'an-GAWB-3100-v2' of github.com:broadinstitute/rawls into an-GAWB-3100-v2

aednichols committed:
Fix compiling

aednichols committed:
Return only Agora MC's from the existing endpoint

Return all MC's from corresponding v2 endpoint

aednichols committed:
Formatting-only changes

aednichols committed:
Disallow posting Dockstore configs to Agora

rushtong commented on core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/MethodConfigurationComponentSpec.scala:
```diff
@@ -81,7 +81,7 @@ class MethodConfigurationComponentSpec extends TestDriverComponentWithFlatSpecAn
       inputs = Map("input.expression.new" -> AttributeString("input.expr")),
       outputs = Map("output.expression.new" -> AttributeString("output.expr")),
-      methodRepoMethod = testData.methodConfig.methodRepoMethod.copy(methodVersion = 2)
+      methodRepoMethod = testData.methodConfig.methodRepoMethod.asInstanceOf[AgoraMethod].copy(methodVersion = 2)
```
Just a thought - can we do this in an `apply` instead of `asInstanceOf` ?

aednichols replied:
We could do
```
AgoraMethod.apply(testData.methodConfig.methodRepoMethod.methodUri).copy(methodVersion = 2)
```
but it would be simpler to just change the test data from
```
val methodConfig = MethodConfiguration(
  [...]
  AgoraMethod("ns-config", "meth1", 1)
)
```
to
```
val method = AgoraMethod("ns-config", "meth1", 1)

val methodConfig = MethodConfiguration(
  [...]
  method
)
```
and use `testData.method.copy(methodVersion = 2)`

aednichols committed:
Simplify test data

aednichols committed:
Clarify

kyuksel commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpMethodRepoDAO.scala:
```diff
@@ -57,8 +58,13 @@ class HttpMethodRepoDAO(baseMethodRepoServiceURL: String, apiPath: String = "",
+      case agoraMethod: AgoraMethod =>
+        getAgoraEntity(s"${methodRepoServiceURL}/methods/${agoraMethod.methodNamespace}/${agoraMethod.methodName}/${agoraMethod.methodVersion}", userInfo)
+      case _ =>
+        throw new RawlsException(s"Method repo \'${method.repo.scheme}\' not yet supported")
```
Nit: `\`s shouldn't be necessary. Same below also.

kyuksel commented on model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala:
```diff
@@ -216,10 +216,14 @@ object MethodRepoMethod {
+
+  def apply(methodPath: String, methodVersion: String): DockstoreMethod = DockstoreMethod(methodPath, methodVersion)
+
 }
```
Add `override` for `def AgoraMethod.validate()`

kyuksel reviewed (commented):
LGTM. Minor comments. Couldn't test.

aednichols committed:
Code review updates

davidangb reviewed (commented):
high-level this looks good; I will continue to review syntax at the low level. I'm trying to understand exactly what is changing in the endpoint/API definitions ... could you document that, probably by updating swagger?

aednichols commented:
Yup, will plug away at the Swagger. I held off for a while because the API itself kept changing with people's input.

aednichols commented on core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/MethodConfigApiService.scala:
```diff
@@ -24,106 +24,113 @@ trait MethodConfigApiService extends HttpService with PerRequestCreator with Use
         }
+      }
+    } ~
+    path("workspaces" / "v2" / Segment / Segment /  "methodconfigs") { (workspaceNamespace, workspaceName) =>
```
We could also call this something like
```
"workspaces" / Segment / Segment /  "allRepoMethodConfigs"
```
which may be more comprehensible to API users.

rushtong replied:
Been mulling this point around. Since we are keeping the original `"workspaces" / Segment / Segment /  "methodconfigs"` endpoint, I'm leaning toward `v2`. We need to have a DSP discussion about versioning semantics and I think this forces our hand. Dockstore is "different enough" &trade; to warrant a new version.

aednichols committed:
Swagger updates

The polymorphism of MethodRepoMethod is pretty hard to coerce into Swagger, probably a few more tweaks needed

aednichols committed:
Swagger updates

aednichols committed:
More Swagger

aednichols committed:
Unit test inspired by writing documentation

aednichols committed:
Documentation

aednichols committed:
Documentation

aednichols committed:
Documentation

rushtong commented:
Some preliminary comments on **POST** functionality. 

* Adam is aware that `methodConfigVersion` and `deleted` are not documented in the swagger and are (currently) required by the api for posting new method configs.
* ~`methodNamespace` is also causing errors for the three uri-based cases:~
```
	"methodRepoMethod": {
		"sourceRepo": "dockstore",
		"methodPath": "broadinstitute/wdl/Validate-Bams",
		"methodVersion": "develop"
	}
	"methodRepoMethod": {
		"methodUri": "dockstore://broadinstitute%2Fwdl%2FValidate-Bams/develop"
	}
	"methodRepoMethod": {
		"methodUri": "agora://adam-methods/test/555"
	}
```
~In those cases, I get a rawls error: `"message": "Object is missing required member 'methodNamespace'"`~

EDIT: The `methodNamespace` is not a problem in rawls, only orch at the moment due to orch not having the updated model classes.

rushtong commented on core/src/main/resources/swagger/rawls.yaml:
```diff
@@ -2111,6 +2158,68 @@ paths:
+          schema:
+            type: array
+            items:
+              $ref: '#/definitions/MethodConfiguration'
```
I think we need `deleted`, `deletedDate`, and `methodConfigVersion` added to `#/definitions/MethodConfiguration`, even if they are optional and have default values.

rushtong commented on core/src/main/scala/org/broadinstitute/dsde/rawls/util/WithSupport.scala:
```diff
@@ -27,9 +27,9 @@ trait MethodWiths {
-  def withMethod[T](methodNamespace: String, methodName: String, methodVersion: Int, userInfo: UserInfo)(op: (AgoraEntity) => ReadWriteAction[T])(implicit executionContext: ExecutionContext): ReadWriteAction[T] = {
-    DBIO.from(methodRepoDAO.getMethod(methodNamespace, methodName, methodVersion, userInfo)).asTry.flatMap {
-      case Success(None) => DBIO.failed(new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.NotFound, s"Cannot get ${methodNamespace}/${methodName}/${methodVersion} from method repo.")))
+  def withMethod[T](method: MethodRepoMethod, userInfo: UserInfo)(op: (AgoraEntity) => ReadWriteAction[T])(implicit executionContext: ExecutionContext): ReadWriteAction[T] = {
```
Nice simplification

rushtong reviewed (commented):
👍 Simultaneous LGTM and needs some changes :-) Definitely need a solution (handle or remove) the fields issue. Otherwise, works as advertised, seems to have adequate tests, and nicely organizes the Agora and Dockstore types.

aednichols committed:
Update MethodConfiguration docs

aednichols committed:
Copy editing

davidangb reviewed (approved):
:+1: for the code (I did not run/test it myself).

As for the API signature, I'm really leaning towards keeping the existing url `/api/workspaces/{workspaceNamespace}/{workspaceName}/methodconfigs`, but adding a query parameter that defaults to agora-only.

aednichols commented:
@rushtong what do you think of the query parameter solution David suggested?

Also, I updated the docs to include those extra fields.

aednichols committed:
Use query param instead

davidangb reviewed (approved):
:+1: banana.

this will need corresponding orch changes - if only to swagger - to match the endpoint change. (edit: and I see you're already on that. Nice!)

aednichols commented:
@davidangb - yes, this is progress as https://github.com/broadinstitute/firecloud-orchestration/pull/607

Because Orch unmarshalls `MethodConfiguration` in the `POST` to check a field or two, Orch will need a Rawls Model update for unmarshalling to work correctly when creating Dockstore configs.

rushtong reviewed (approved):
👍

aednichols committed:
Undo formatting changes to reduce conflicts
